northstar/src/runtime/state.rs

```diff
@@ -109,617 +95,494 @@ impl ProcessContext {
+            let (dir, repository) = prepare_internal_repository(&name).await?;
             repositories.insert(name, repository);
             dir
         };
 
+        // Build a map of repositories from the configuration
         for (id, repository) in &config.repositories {
-            let key = {
-                if let Some(key) = repository.key.as_ref() {
-                    Some(keys::load(&key).await.map_err(Error::Key)?)
-                } else {
-                    None
-                }
-            };
-            let dir = repository.dir.clone();
             repositories.insert(
                 id.clone(),
-                Repository {
-                    id: id.clone(),
-                    dir,
-                    key,
-                },
+                Repository::new(
+                    id.clone(),
+                    repository.dir.clone(),
+                    repository.key.as_deref(),
+                )
+                .await?,
             );
         }
 
-        let minijail = Minijail::new(tx.clone(), config)
+        // TODO: Verify that the containers in all repositories are unique with name and version
+
+        let minijail = Minijail::new(events_tx.clone(), config)
             .await
             .map_err(Error::Process)?;
+        let mount_control = MountControl::new(&config).await.map_err(Error::Mount)?;
 
-        let mut state = State {
-            events_tx: tx,
+        Ok(State {
+            events_tx,
             repositories,
-            applications: HashMap::new(),
-            resources: HashMap::new(),
+            containers: HashMap::new(),
             config,
             minijail,
-            mount_control: MountControl::new(&config).await.map_err(Error::Mount)?,
-            #[cfg(debug_assertions)]
+            mount_control: Arc::new(mount_control),
+            #[cfg(feature = "hello-world")]
             internal_repository,
-        };
-
-        // Mount all the containers from each repository
-        state.mount_repositories().await?;
-        Ok(state)
+        })
     }
 
-    /// Return an iterator over all known applications
-    pub fn applications(&self) -> impl iter::Iterator<Item = &Application> {
-        self.applications.values()
+    fn npk(&self, container: &Container) -> Option<(&Path, &Npk, Option<&PublicKey>)> {
+        for repository in self.repositories.values() {
+            if let Some(path) = repository.containers.get(container) {
+                let npk = repository.npks.get(container)?;
+                return Some((path, npk, repository.key.as_ref()));
+            }
+        }
+        None
     }
 
-    /// Return an iterator over all known resources
-    pub fn resources(&self) -> impl iter::Iterator<Item = &Container> {
-        self.resources.values()
+    /// Mount `container`
+    async fn mount(
+        &self,
+        container: &Container,
+    ) -> Result<impl Future<Output = Result<MountedContainer, Error>>, Error> {
```

Comment:
Just as before I don't understand the use for the double future.

Reply:
If `Npk::from_path` would not be async this would not be needed.